### PR TITLE
Support node pools CRUD

### PR DESCRIPTION
**Pull request by YevheniiSemendiak**

### CODE REVIEW

File: neuro_config_client/client.py [hunk 1]
```diff
@@ -276,6 +279,159 @@ async def remove_storage(
...
+            return [self._entity_factory.create_node_pool(n) for n in resp_payload]
+
+    async def add_node_pool(
+        self,
+        cluster_name: str,
+        node_pool: NodePool,
+        *,
+        token: str | None = None,
+        start_deployment: bool = True,
+    ) -> Cluster:
```
Comment (YevheniiSemendiak) [inline neuro_config_client/client.py:326]: In the future, we might need to change the API so it returns added/patched/deleted node_pool. WDYT?

File: neuro_config_client/factories.py [hunk 2]
```diff
@@ -308,6 +308,7 @@ def _create_node_pool(self, payload: dict[str, Any]) -> NodePool:
             machine_type=payload.get("machine_type"),
             idle_size=payload.get("idle_size", NodePool.idle_size),
             is_preemptible=payload.get("is_preemptible", NodePool.is_preemptible),
+            zones=payload.get("zones", ()),
```
Comment (YevheniiSemendiak) [inline neuro_config_client/factories.py:311]: `zones` info in NodePool is a legacy?
I see it in ClusterProvider.

File: neuro_config_client/client.py [hunk 3]
```diff
@@ -276,6 +279,159 @@ async def remove_storage(
...
+        url = (
+            self._clusters_url
+            / cluster_name
+            / "cloud_prodider"
+            / "node_pools"
+            / node_pool_name
+        )
+
+        headers = self._create_headers(token=token)
+        async with self._client.get(url=url, headers=headers) as responce:
```
Comment (zubenkoivan) [inline neuro_config_client/client.py:299]: typo, response

File: neuro_config_client/client.py [hunk 4]
```diff
@@ -276,6 +279,159 @@ async def remove_storage(
...
+        async with self._client.post(
+            url.with_query(start_deployment=str(start_deployment).lower()),
+            headers=headers,
+            json=payload,
+        ) as response:
+            response.raise_for_status()
+            resp_payload = await response.json()
+            return self._entity_factory.create_cluster(resp_payload)
+
+    async def patch_node_pool(
```
Comment (zubenkoivan) [inline neuro_config_client/client.py:342]: Why did you add `patch_node_pool` method? We don't have patch in api.

File: neuro_config_client/factories.py (hunk 2 above)
Comment (zubenkoivan) [reply neuro_config_client/factories.py:311]: It's not legacy, it is used

File: neuro_config_client/factories.py [hunk 5]
```diff
@@ -814,3 +815,34 @@ def create_ingress(cls, ingress: IngressConfig) -> dict[str, Any]:
         if ingress.cors_origins:
             result["cors_origins"] = ingress.cors_origins
         return result
+
+    @classmethod
+    def create_node_pool(cls, node_pool: NodePool) -> dict[str, Any]:
```
Comment (zubenkoivan) [inline neuro_config_client/factories.py:820]: This payload cannot be used for adding/updating node pools. Payload depends on the type of cluster cloud provider: for cloud clusters machine configuration is taken from node templates which are stored on the server, for onprem clusters machine configuration is provided by client

File: neuro_config_client/client.py (hunk 1 above)
Comment (zubenkoivan) [reply neuro_config_client/client.py:326]: When you change node pools other parts of cluster also change, so it's better to return the whole cluster. The same behavior for storages.

File: neuro_config_client/client.py [hunk 6]
```diff
@@ -276,6 +279,159 @@ async def remove_storage(
...
+        headers = self._create_headers(token=token)
+        async with self._client.get(url=url, headers=headers) as responce:
+            responce.raise_for_status()
+            resp_payload = await responce.json()
+            return [self._entity_factory.create_node_pool(n) for n in resp_payload]
+
+    async def add_node_pool(
+        self,
+        cluster_name: str,
+        node_pool: NodePool,
```
Comment (zubenkoivan) [inline neuro_config_client/client.py:322]: You cannot use this model to add/update node pool. It's different for cloud and onprem clusters.

File: neuro_config_client/client.py (hunk 3 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/client.py]: fixed in [4b49d1a](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/4b49d1afae1854dfa25aa96e86fa5cbd1bed6224)

File: neuro_config_client/factories.py (hunk 5 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/factories.py]: Good catch, thank you for help! checkout [be58386](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/be5838666c221fd64f7f1085e93d2598905e231a)

File: neuro_config_client/client.py (hunk 6 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/client.py:345]: same here, thank you for help, pls check [be58386](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/be5838666c221fd64f7f1085e93d2598905e231a)

File: neuro_config_client/entities.py [hunk 7]
```diff
@@ -61,6 +61,49 @@ class NodePool:
...
+            cpu=self.cpu,
+            available_cpu=self.available_cpu,
+            memory_mb=self.memory_mb,
+            available_memory_mb=self.available_memory_mb,
+            gpu=self.gpu,
+            gpu_model=self.gpu_model,
+        )
+
+
+def get_node_pool_id(machine_type: str, cpu: float) -> str:
```
Comment (zubenkoivan) [inline neuro_config_client/entities.py:98]: please remove it, client should receive id from server

File: neuro_config_client/factories.py [hunk 8]
```diff
@@ -814,3 +828,92 @@ def create_ingress(cls, ingress: IngressConfig) -> dict[str, Any]:
...
+        if node_pool.disk_type:
+            result["disk_type"] = node_pool.disk_type
+        if node_pool.gpu:
+            result["gpu"] = node_pool.gpu
+        if node_pool.gpu_model:
+            result["gpu_model"] = node_pool.gpu_model
+        if node_pool.currency:
+            result["currency"] = node_pool.currency
+        if node_pool.zones:
+            result["zones"] = node_pool.zones
```
Comment (zubenkoivan) [inline neuro_config_client/factories.py]: Server validator doesn't accept zones

File: neuro_config_client/factories.py [hunk 9]
```diff
@@ -814,3 +828,92 @@ def create_ingress(cls, ingress: IngressConfig) -> dict[str, Any]:
...
+            "available_cpu": node_pool.available_cpu,
+            "memory_mb": node_pool.memory_mb,
+            "available_memory_mb": node_pool.available_memory_mb,
+            "disk_size_gb": node_pool.disk_size_gb,
+            "price": str(node_pool.price),
+            "is_preemptible": node_pool.is_preemptible,
+            "zones": node_pool.zones,
+        }
+        if node_pool.machine_type:
+            result["machine_type"] = node_pool.machine_type
```
[the remark this answers is not in the document]
Comment (YevheniiSemendiak) [reply neuro_config_client/factories.py]: removed in [1972938](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/1972938fc1d597d4fb0cbd449776a344b17e53a2)

File: neuro_config_client/factories.py (hunk 8 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/factories.py]: removed in [1972938](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/1972938fc1d597d4fb0cbd449776a344b17e53a2)

File: neuro_config_client/entities.py (hunk 7 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/entities.py]: done in [44c6d27](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/44c6d279d91f77f339668d2979e94a735354c35d)

File: neuro_config_client/factories.py [hunk 10]
```diff
@@ -814,3 +828,92 @@ def create_ingress(cls, ingress: IngressConfig) -> dict[str, Any]:
...
             result["cors_origins"] = ingress.cors_origins
         return result
+
+    @classmethod
+    def create_node_pool(
+        cls, node_pool: NodePool, cloud_provider_type: CloudProviderType
+    ) -> dict[str, Any]:
+        if cloud_provider_type == CloudProviderType.ON_PREM:
+            return cls._create_onprem_node_pool(node_pool)
+        elif cloud_provider_type.value.startswith("vcd_"):
```
[the remark this answers is not in the document]
Comment (YevheniiSemendiak) [reply neuro_config_client/factories.py:840]: I was thinking to do so, but now have doubts about it, since the payload differs and is validated differently. For instance:
- idle_size, is_preemptible, zones are allowed in aws/gcp/azure but banned in vcd
- price and currency are banned in aws/gcp/azure, but allowed in vcd
- disk_type is optional for aws/gcp/azure but required for vcd

A single could be overcomplicated with if-else and, basically, we will have this separate method under other `elif cloud_provider_type.value.startswith("vcd_")`
I guess, thereby we validate payload separately on the server side.

wdyt?

File: neuro_config_client/factories.py [hunk 11]
```diff
@@ -814,3 +830,85 @@ def create_ingress(cls, ingress: IngressConfig) -> dict[str, Any]:
...
+            "disk_size_gb": node_pool.disk_size_gb,
+        }
+        if node_pool.idle_size:
+            result["idle_size"] = node_pool.idle_size
+        if node_pool.is_preemptible:
+            result["is_preemptible"] = node_pool.is_preemptible
+        if node_pool.disk_type:
+            result["disk_type"] = node_pool.disk_type
+
+        if cloud_provider_type == CloudProviderType.AWS:
```
Comment (zubenkoivan) [inline neuro_config_client/factories.py:888]: What if developer passes node_pool.gpu / gpu_model for cpu node? I think he should receive HTTP 400 error from server, but in your case he receives HTTP 200.

File: neuro_config_client/client.py [hunk 12]
```diff
@@ -276,6 +283,299 @@ async def remove_storage(
...
+        *,
+        token: str | None = None,
+        start_deployment: bool = True,
+    ) -> Cluster:
+        assert self._client
+        assert cluster.cloud_provider
+        assert node_pool.id
+
+        np_templates = await self.get_node_pool_templates(
```
Comment (zubenkoivan) [inline neuro_config_client/client.py:424]: This request is redundant, it's server responsibility to check that id/templates exist. Developer should pass correct id to the client. If developer needs to do these checks on the client side let him do it himself.
When I see function `add_node_pool` I don't expect it to do multiple requests on server, make validation on client. I expect it to make only one request to server with the data I provided and let the server decide if the data is correct.

File: neuro_config_client/factories.py (hunk 11 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/factories.py:888]: Do you mean, we should pass zones, gpu and gpu_model info if they are present for all cases?

For me, as from an end-user perspective, this is mostly the same as we also pass cpu/memory info.
It's not clear which info should be included in the node pool description, especially since the HTTP error response from the server does not include info regarding what is wrong in the node pool description (validation results are not included).

I could include zones, gpu, gpu_model to all three types, but we should also send back the validation results in case of error. WDYT?

Comment (zubenkoivan) [reply neuro_config_client/factories.py:888]: Yes, pass all the data. Validation results is another problem, it needs to be tackled separately for all the client requests.

Comment (YevheniiSemendiak) [reply neuro_config_client/factories.py]: Done in d229e1f

File: neuro_config_client/client.py (hunk 12 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/client.py]: That makes sense, thanks for the clarification! Removed those steps in [d229e1f](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/d229e1fabcc31e90a63c8084af36f6024a7e6780)

File: neuro_config_client/client.py [hunk 13]
```diff
@@ -276,6 +279,159 @@ async def remove_storage(
...
+        cluster_name: str,
+        node_pool_name: str,
+        *,
+        token: str | None = None,
+    ) -> NodePool:
+        assert self._client
+        url = (
+            self._clusters_url
+            / cluster_name
+            / "cloud_prodider"
```
Comment (zubenkoivan) [inline neuro_config_client/client.py]: cloud_provider

File: neuro_config_client/client.py [hunk 14]
```diff
@@ -276,6 +282,207 @@ async def remove_storage(
...
+        token: str | None = None,
+        start_deployment: bool = True,
+    ) -> Cluster:
+        assert self._client
+
+        current_np = await self.get_node_pool(cluster_name, node_pool_name, token=token)
+        cluster = await self.get_cluster(cluster_name, token=token)
+        assert cluster.cloud_provider
+
+        new_np = NodePool(
```
Comment (zubenkoivan) [inline neuro_config_client/client.py:418]: This won't work, server validation will fail. If we need to patch node pools then let's add PATCH endpoint.

File: neuro_config_client/entities.py [hunk 15]
```diff
@@ -62,6 +63,31 @@ class NodePool:
...
+    id: str
+    machine_type: str
+    cpu: float
+    available_cpu: float
+    memory_mb: int
+    available_memory_mb: int
+    gpu: int | None = None
+    gpu_model: str | None = None
+
+    def to_nodepool(self, name: str | None = None) -> NodePool:
```
[the remark this answers is not in the document]
Comment (YevheniiSemendiak) [reply neuro_config_client/entities.py]: I was used it with the previous implementation, but now the only need is the `id`, so I removed it in [ed0f260](https://github.com/neuro-inc/neuro-config-client/pull/83/commits/ed0f26067f2b3c31d5e3ea29c7dc8e4cca07d3e3)

File: neuro_config_client/client.py (hunk 14 above)
Comment (YevheniiSemendiak) [reply neuro_config_client/client.py]: It worked when I cut off unneeded info at the payload factory, but you are right.
Added feature request, https://github.com/neuro-inc/platform-config/issues/1276, updated method.

**Review by zubenkoivan — APPROVED** [review]
Thanks! Good job! After release please upgrade client in `platform-config` and add some integration tests
